### PR TITLE
Poll git repo

### PR DESCRIPTION
**Pull request by jssblck**
# Overview

Adds git repo polling to the `run` subcommand, and puts new references into the async queue to be processed.

## Acceptance criteria

New references are enqueued for processing.

## Testing plan

This is mostly just gluing together existing functionality, so no new tests.
I did run this in my terminal:
```
; make run
   Compiling broker v0.1.0-pre (/Users/jessica/projects/broker)
    Finished dev [unoptimized + debuginfo] target(s) in 2.69s
     Running `target/debug/broker run`
  2023-03-18T00:01:54.035554Z  INFO broker::debug: Debug artifacts being stored in "/Users/jessica/.config/fossa/broker/debug/trace"
    at src/debug.rs:116

  2023-03-18T00:01:54.045007Z  INFO broker: Loaded Config { fossa_api: Config { endpoint: Endpoint(Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("app.fossa.com")), port: None, path: "/", query: None, fragment: None }), key: Key(ComparableSecret(REDACTED)) }, debug: Config { location: Root("/Users/jessica/.config/fossa/broker/debug"), retention: Retention { days: ArtifactRetentionCount(7) } }, integrations: Integrations([Integration { poll_interval: PollInterval(3600s), protocol: Git(Http { endpoint: Remote("https://github.com/fossas/fossa-cli.git"), auth: None }) }]) }
    at src/main.rs:136

  2023-03-18T00:02:05.249746Z  INFO broker::subcommand::run: Enqueued task to scan using git protocol with transport cloning via HTTP from https://github.com/fossas/fossa-cli.git with no authentication at git branch 'ANE-891-dynamic-go-tactic' with head commit 'af759ce81175b748cb8f1f2f8899baf6e1677d71'
    at src/subcommand/run.rs:244
    in broker::subcommand::run::do_poll_integration
    in broker::subcommand::run::do_poll_integrations
    in broker::subcommand::run::main

... some more tasks ...

  2023-03-18T00:02:05.250517Z  INFO broker::subcommand::run: Next poll interval for using git protocol with transport cloning via HTTP from https://github.com/fossas/fossa-cli.git with no authentication in 3600s
    at src/subcommand/run.rs:263
    in broker::subcommand::run::do_poll_integration
    in broker::subcommand::run::do_poll_integrations
    in broker::subcommand::run::main

  2023-03-18T00:02:05.250790Z  INFO broker::subcommand::run: Pretending to scan using git protocol with transport cloning via HTTP from https://github.com/fossas/fossa-cli.git with no authentication at git branch 'ANE-891-dynamic-go-tactic' with head commit 'af759ce81175b748cb8f1f2f8899baf6e1677d71'
    at src/subcommand/run.rs:284
    in broker::subcommand::run::do_scan_git_reference with job: ScanGitVCSReference { scan_id: "356877f4-7d21-4803-8623-549466292f19", integration: Integration { poll_interval: PollInterval(3600s), protocol: Git(Http { endpoint: Remote("https://github.com/fossas/fossa-cli.git"), auth: None }) }, reference: Git(Branch { name: "ANE-891-dynamic-go-tactic", head: "af759ce81175b748cb8f1f2f8899baf6e1677d71" }) }, scan_id: "356877f4-7d21-4803-8623-549466292f19"
    in broker::subcommand::run::do_scan_git_references
    in broker::subcommand::run::main

... some more tasks ...
```

This did expose some issues:
- ctrl+c leaves the task queues in an "open" state, forcing manual deletion of their lockfiles.
- I can't get it to clone SSH repos, this may be a bug in the git library.
- The code to keep this functionality generalized is starting to spaghettify pretty badly, if I get time I may take a redesign pass on how we organize our types & modules
- The wording in these info messages is _really_ verbose, I'll probably take a pass at displaying these in a more useful way to the user.
- Working with async for this problem is a bit on the complicated side, I think it may be more worthwhile to use plain old threads, but I stuck with async for this PR as that's what we've done so far.

I want to keep this PR on the smaller side, so I'd like to merge despite these issues and address them in future PRs.

## References

Implements https://fossa.atlassian.net/browse/ANE-837

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).

### CODE REVIEW

File: src/ext/io.rs [hunk 1]
```diff
@@ -144,5 +144,17 @@ where
     Report<E>: From<E>,
     F: FnOnce() -> Result<T, E> + Send + 'static,
 {
-    run_background(|| work().into_report()).await
+    spawn_blocking_stacked(|| work().into_report()).await
+}
+
+/// Run the provided blocking closure in the background,
```
Comment (spatten) [inline src/ext/io.rs:150]: I don't understand when I would use `spawn_blocking` and when I would use `spawn_blocking_stacked`, and their comments are identical.  Can we make that more clear?

File: src/subcommand/run.rs [hunk 2]
```diff
@@ -21,47 +37,253 @@ pub enum Error {
...
+            Ok(success) => Ok(success),
+            Err(err) => {
+                warn!("attempt to poll integration at {remote} failed: {err:#}");
+                Err(err)
+            }
+        }
+    };
+
+    loop {
+        // Given that this operation is not latency sensitive, and temporary neetwork issues can interfere,
```
Comment (spatten) [inline src/subcommand/run.rs:177]: Spelling nit:

```suggestion
        // Given that this operation is not latency sensitive, and temporary network issues can interfere,
```

File: src/subcommand/run.rs [hunk 3]
```diff
@@ -21,47 +37,253 @@ pub enum Error {
...
+        for reference in references {
+            let job = ScanGitVCSReference::new(integration, &reference);
+            let encoded = bincode::serialize(&job).context(Error::TaskEncode)?;
+            let mut sender = sender.lock().await;
+            sender
+                .send(&encoded)
+                .await
+                .change_context(Error::TaskEnqueue)?;
+
+            info!("Enqueued task to scan {integration} at {reference}");
```
Comment (spatten) [inline src/subcommand/run.rs:244]: I found these a bit hard to parse. Here's what they currently look like:

```
Enqueued task to scan using git protocol with transport cloning via SSH from git@github.com:fossas/broker.git with key file at "/Users/scott/.ssh/id_ed25519" at git branch 'test-with-async_with_vars' with head commit '98f35745378eaaf375b0df6e055e7166077c677f'
```

I tried two things to see if I could improve this, neither of which is perfect.

First, I just wrapped some phrases in quotes to group them:

```
Enqueued task "scan using git protocol" with transport "cloning via SSH" from git@github.com:fossas/broker.git with key file at "/Users/scott/.ssh/id_ed25519" at git branch 'test-with-async_with_vars' with head commit '98f35745378eaaf375b0df6e055e7166077c677f'
```

I also tried making things more "JSONy" like this, but this might be deep in the uncanny valley:

```
Enqueuing task: "scan using git protocol", transport: "cloning via SSH", remote: git@github.com:fossas/broker.git, auth: "key file at '/Users/scott/.ssh/id_ed25519'", reference: "git branch 'test-with-async_with_vars' with head commit '98f35745378eaaf375b0df6e055e7166077c677f'"
```

**Review by spatten — APPROVED** [review]
Looks great! I added a few comments, but they are not at all blocking

File: src/subcommand/run.rs (hunk 3 above)
Comment (jssblck) [reply src/subcommand/run.rs:244]: yeah, this is the "the wording in these info messages is really verbose, I'll probably take a pass at displaying these in a more useful way to the user".

I'd like to leave it here for now but with the knowledge/hope that we'll go touch these up. I am not yet certain how to make them useful for the user but without making them literal json-style outputs (which semi-technical users will find hard to parse as well). I suspect that if we address the "the code to keep this functionality generalized is starting to spaghettify pretty badly, if I get time I may take a redesign pass on how we organize our types & modules" issue this may get easier as well.

File: src/ext/io.rs (hunk 1 above)
Comment (jssblck) [reply src/ext/io.rs:150]: The difference is in the types: `spawn_blocking` changes the error to an `error_stack::Report`, while `spawn_blocking_stacked` expects an existing `error_stack::Report`.

I'll definitely disambiguate these before merging.
